src/server/parser.rs

```diff
@@ -317,144 +123,212 @@ fn consumed(off: usize) -> usize {
     if off > 0 { off+2 } else { 0 }
 }
 
-impl<M: Server> ParserImpl<M>
-{
-    fn wrap<S: StreamSocket>(self) -> Parser<M, S>
-    {
-        Parser(self, PhantomData)
+#[derive(Debug)]
+pub enum Parser<M: Server, S: StreamSocket> {
+    Idle,
+    ReadHeaders,
+    ReadingBody(ReadBody<M>),
+    Processing(M, MessageState, bool, Time),
+    DoneResponse,
+    #[doc(hidden)]
+    __Hidden(PhantomData<*const S>),
```

Comment:
I think it may be useful to do `__Hidden(Void, PhantomData)` and
```
use rotor::void::{Void, unreachable};
match self {
  __Hidden(v, _) => unreachable(v),
}
```
I think this should hint the compiler to completely eliminate enum variant rather than insert panics everywhere.